edgeless_node/src/resources/ollama.rs

```diff
@@ -0,0 +1,344 @@
+// SPDX-FileCopyrightText: © 2023 Claudio Cicconetti <c.cicconetti@iit.cnr.it>
+// SPDX-License-Identifier: MIT
+
+use futures::{SinkExt, StreamExt};
+
+#[derive(Clone)]
+pub struct OllamaResourceProvider {
+    inner: std::sync::Arc<tokio::sync::Mutex<OllamaResourceProviderInner>>,
+}
+
+enum OllamaCommand {
+    // model_name, history_id, prompt, resource_id, reply_receiver
+    Chat(
+        String,
+        String,
+        String,
+        edgeless_api::function_instance::ComponentId,
+        tokio::sync::oneshot::Sender<anyhow::Result<(edgeless_api::function_instance::InstanceId, String)>>,
+    ),
+    // resource_id, target
+    Patch(edgeless_api::function_instance::ComponentId, edgeless_api::function_instance::InstanceId),
+}
+
+impl std::fmt::Display for OllamaCommand {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            OllamaCommand::Chat(model_name, history_id, prompt, resource_id, _) => write!(
+                f,
+                "model {}, history_id {}, prompt length {}, resource_id {})",
+                model_name,
+                history_id,
+                prompt.len(),
+                resource_id
+            ),
+            OllamaCommand::Patch(resource_id, target) => write!(f, "resource_id {}, target {}", resource_id, target),
+        }
+    }
+}
+
+pub struct OllamaResourceProviderInner {
+    resource_provider_id: edgeless_api::function_instance::InstanceId,
+    dataplane_provider: edgeless_dataplane::handle::DataplaneProvider,
+    instances: std::collections::HashMap<edgeless_api::function_instance::ComponentId, OllamaResource>,
+    sender: futures::channel::mpsc::UnboundedSender<OllamaCommand>,
+    _handle: tokio::task::JoinHandle<()>,
+}
+
+pub struct OllamaResource {
+    join_handle: tokio::task::JoinHandle<()>,
+}
+
+impl Drop for OllamaResource {
+    fn drop(&mut self) {
+        self.join_handle.abort();
+    }
+}
+
+impl OllamaResource {
+    /// Create a new Ollama resource.
+    ///
+    /// - `resource_provider`: the resource provider.
+    /// - `dataplane_handle`: gives access to the EDGELESS dataplane.
+    /// - `model_name`: name of the AI model to use.
+    /// - `instance_id`: identifier of this resource instance.
+    /// - `sender`: channel to send commands to the resource task.
+    async fn new(
+        dataplane_handle: edgeless_dataplane::handle::DataplaneHandle,
+        model_name: String,
+        instance_id: edgeless_api::function_instance::InstanceId,
+        sender: futures::channel::mpsc::UnboundedSender<OllamaCommand>,
+    ) -> anyhow::Result<Self> {
+        let mut dataplane_handle = dataplane_handle;
+        let mut sender = sender;
+
+        let history_id = instance_id.function_id.to_string();
+        let handle = tokio::spawn(async move {
```

Comment:
Would keep those out of the new / spawn method and create a struct to hold the state.
This is rather hard to read, especially with the second loop in the provider.